gc.c

```diff
@@ -3304,30 +3318,57 @@ incremental_enable(VALUE _)
 void
 rb_objspace_each_objects(each_obj_callback *callback, void *data)
 {
-    objspace_each_objects(&rb_objspace, callback, data);
+    objspace_each_objects(&rb_objspace, callback, data, TRUE);
 }
 
 static void
-objspace_each_objects(rb_objspace_t *objspace, each_obj_callback *callback, void *data)
+objspace_each_objects(rb_objspace_t *objspace, each_obj_callback *callback, void *data, bool protected)
 {
-    int prev_dont_incremental = objspace->flags.dont_incremental;
+    /* Disable incremental GC */
+    bool reenable_incremental = FALSE;
+    if (protected) {
+        reenable_incremental = !objspace->flags.dont_incremental;
 
-    gc_rest(objspace);
-    objspace->flags.dont_incremental = TRUE;
-
-    if (prev_dont_incremental) {
-        objspace_each_objects_without_setup(objspace, callback, data);
+        gc_rest(objspace);
+        objspace->flags.dont_incremental = TRUE;
     }
-    else {
-        struct each_obj_args args = {objspace, callback, data};
-        rb_ensure(objspace_each_objects_protected, (VALUE)&args, incremental_enable, Qnil);
+
+    /* Create pages buffer */
+    struct heap_page **pages = malloc(sizeof(struct heap_page *) * heap_allocated_pages);
```

Comment:
I think we should use `size_mul_or_raise` here. Like [this](https://github.com/ruby/ruby/blob/061edf6b051e1e10f388fa92f9ae4f161008864e/gc.c#L1656)